src/box/lua/net_cluster.lua

```diff
@@ -0,0 +1,466 @@
+    return watcher
+end
+
+-- Trigger internal load of config (in case of `config_mode` == 'auto')
+-- set new cluster config (in case of `config_mode` == 'manual')
+-- In the first case `config` must be nil, in the second - it must be
+-- a map replicaset_name -> replicaset_cfg (table).
+-- replicaset_cfg may have members:
+--  connect_cfg: if provided, it will be passed to net.replicaset.connect.
+--   (if not, `replicaset_name` will be passed along with connect_opts).
+--  roles: optional list of roles (map or array) that the replicaset have.
+--   if present, affects filtering by role in some methods.
+function cluster_methods:reload_config(config)
+    cluster_check(self, 'reload_config')
+    if config then
+        if self._cfg.config_mode ~= 'manual' then
+            box.error(box.error.ILLEGAL_PARAMS,
+                    'manual reload expects manual config mode')
+        end
+        reload_config_impl(self, config)
+    else
+        if self._cfg.config_mode ~= 'auto' then
+            box.error(box.error.ILLEGAL_PARAMS,
+                    'force reload expects auto config mode')
+        end
+        reload_config_impl(self, collect_config())
+    end
+end
+
+-- Close all watchers and connections.
+function cluster_methods:close()
+    cluster_check(self, 'close')
+    while true do
+        local watcher = next(self._watchers)
+        if not watcher then
+            break
+        end
+        watcher:unregister()
+    end
+    for _, replicaset in pairs(self._replicasets) do
+        close_replicaset_connection(replicaset)
+    end
+    table.clear(self._replicasets)
+    if self._reload_watcher then
+        self._reload_watcher:unregister()
+        self._reload_watcher = nil
+    end
+    self._is_closed = true
+end
+
+-- get for default cluster (which is created if needed).
+local function get(replicaset_name)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:get(replicaset_name)
+end
+
+-- get_connections for default cluster (which is created if needed).
+local function get_connections(opts)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:get_connections(opts)
+end
+
+-- watch_leaders for default cluster (which is created if needed).
+local function watch_leaders(key, func, opts)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:watch_leaders(key, func, opts)
+end
+
+-- reload_config for default cluster (which is created if needed).
+local function reload_config(config)
+    if not default_cluster then
+        default_cluster = cluster_new(cluster_cfg_defaults)
+    end
+    return default_cluster:reload_config(config)
+end
+
+return {
+    new = cluster_new,
+
+    get = get,
+    get_connections = get_connections,
+    watch_leaders = watch_leaders,
```

Comment:
In my understanding, the publish-subscribe pattern requires the ability to set handlers for the following situations:

* on connect
* on change
* on disconnect

I'm thinking on it in terms of a local represention/a local caching of a remote state. This local cache has the `not yet filled` state, has the `actual` state and has the `stalled/expired` state. We need the 'on disconnect' handler to implement the `stalled/expired` state.

The 'on connect' handler is coalesced with 'on change' in `net.box`, it is OK for my taste.

We can call 'on change' periodically even if the remote state is not changed and don't provide 'on disconnect'. However, `net.box` provides the `on_disconnect` trigger separately (and don't call 'on change' if there were no changes). Both ways are OK in my understanding.

Note re `net.box`: it may be used to actually track actuality of the local state after #9563.